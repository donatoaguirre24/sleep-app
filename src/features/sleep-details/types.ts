export type SubjectSleepData = {
  intervals: Interval[]
}

export type Interval = {
  id: string
  ts: Date
  stages: Array<StageElement>
  score: number
  timeseries: TimeSeries
}

export type StageElement = {
  stage: StageEnum
  duration: number
}

export enum StageEnum {
  Awake = 'awake',
  Deep = 'deep',
  Light = 'light',
  Out = 'out',
}

export type TimeSeries = {
  tnt: Array<Array<Date | number>>
  tempRoomC: Array<Array<Date | number>>
  tempBedC: Array<Array<Date | number>>
  respiratoryRate: Array<Array<Date | number>>
  heartRate: Array<Array<Date | number>>
  heating: Array<Array<Date | number>>
}

export type TemperatureSeries = Array<Array<Date | number>>
