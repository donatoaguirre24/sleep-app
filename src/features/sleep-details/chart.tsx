import React from 'react'
import { SafeAreaView } from 'react-native'

import { Card, Text } from '@/components'
import { SleepDetailsRouteProp } from '@/navigation'

type Props = {
  route: SleepDetailsRouteProp
}

export const Chart: React.FC<Props> = ({ route }) => {
  const intervalId = route.params.intervalId

  return (
    <SafeAreaView>
      <Card backgroundColor="green" borderRadius={5} margin="m" height={200} />
      <Text textAlign="center">{`Sleep data for interval: ${intervalId}`}</Text>
    </SafeAreaView>
  )
}
