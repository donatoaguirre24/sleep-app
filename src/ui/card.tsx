import { createRestyleComponent, createVariant, VariantProps } from '@shopify/restyle'

import { Box } from './box'
import { Theme } from './theme'

export type CardProps = VariantProps<Theme, 'cardVariants'> & React.ComponentProps<typeof Box>

export const Card = createRestyleComponent<CardProps, Theme>(
  [createVariant({ themeKey: 'cardVariants' })],
  Box
)
