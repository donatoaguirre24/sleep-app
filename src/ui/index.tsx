export * from './box'
export * from './card'
export * from './circular-progress'
export * from './screen'
export * from './text'
export * from './theme'
