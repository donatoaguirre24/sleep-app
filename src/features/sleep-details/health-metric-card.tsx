import { ResponsiveValue } from '@shopify/restyle'
import React from 'react'

import { Box, Card, CardProps, Text, Theme } from '@/ui'

type Props = CardProps & {
  title: string
  metric: string | number
  unit: string
  color: ResponsiveValue<keyof Theme['colors'], Theme>
}

export const HealthMetricCard: React.FC<Props> = ({ title, metric, unit, color, ...props }) => {
  return (
    <Card backgroundColor="card" minHeight={75} variant="elevated" {...props}>
      <Text variant="card" color={color}>
        {title}
      </Text>
      <Box flexDirection="row" alignItems="center" flex={1}>
        <Text variant="header" color="text">
          {metric}{' '}
          <Text variant="card" color="secondaryGray">
            {unit}
          </Text>
        </Text>
      </Box>
    </Card>
  )
}
