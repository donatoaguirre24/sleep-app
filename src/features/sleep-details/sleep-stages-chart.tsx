import { useTheme } from '@shopify/restyle'
import React from 'react'
import {
  VictoryArea,
  VictoryAxis,
  VictoryChart,
  VictoryScatter,
  VictoryTheme,
} from 'victory-native'

import { datePlusSeconds, hourStringFromISODate } from '@/services'
import { Theme } from '@/ui'
import { Interval, StageElement, StageEnum } from './types'

type Props = {
  interval: Interval
}

export const SleepStagesChart: React.FC<Props> = ({ interval }) => {
  const { colors } = useTheme<Theme>()

  const data = formatStages(interval.stages, interval.ts)

  const computeColorFromStage = (stage: StageEnum) => {
    switch (stage) {
      case StageEnum.Out:
        return colors.orange
      case StageEnum.Awake:
        return colors.purple
      case StageEnum.Light:
        return colors.green
      case StageEnum.Deep:
        return colors.blue
    }
  }

  return (
    <VictoryChart
      animate={{ duration: 500 }}
      height={250}
      padding={{ top: 20, bottom: 30, left: 50, right: 50 }}
      scale={{ x: 'time', y: 'linear' }}
      theme={VictoryTheme.material}
    >
      <VictoryAxis
        dependentAxis
        categories={{ y: [StageEnum.Deep, StageEnum.Light, StageEnum.Awake, StageEnum.Out] }}
        style={{
          axis: {
            stroke: colors.text,
          },
          grid: {
            stroke: 'none',
          },
          ticks: {
            stroke: 'none',
          },
          tickLabels: {
            fill: ({ text: stage }: { text: StageEnum }) => computeColorFromStage(stage),
          },
        }}
      />
      <VictoryAxis
        fixLabelOverlap
        style={{
          axis: {
            stroke: colors.text,
          },
          grid: {
            stroke: 'none',
          },
          ticks: {
            stroke: 'none',
          },
          tickLabels: {
            fill: colors.text,
          },
        }}
        tickFormat={(date: Date) => hourStringFromISODate(date)}
      />
      <VictoryArea
        data={data}
        interpolation="stepAfter"
        style={{
          data: {
            fill: colors.gray,
            fillOpacity: 0.25,
            stroke: colors.gray,
          },
        }}
      />
      <VictoryScatter
        animate={{
          onLoad: {
            before: () => ({ opacity: 0 }),
            after: () => ({ opacity: 1 }),
          },
        }}
        data={data}
        style={{
          data: {
            fill: ({ datum }) => computeColorFromStage(datum.y),
            opacity: ({ datum }) => datum.opacity,
          },
        }}
      />
    </VictoryChart>
  )
}

function formatStages(stages: Array<StageElement>, intervalStart: Date) {
  let delta = 0

  return stages.map(({ stage, duration }) => {
    const stageStart = datePlusSeconds(intervalStart, delta)

    delta += duration

    return {
      x: stageStart,
      y: stage,
    }
  })
}
