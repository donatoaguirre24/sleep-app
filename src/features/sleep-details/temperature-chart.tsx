import { useTheme } from '@shopify/restyle'
import React from 'react'
import {
  VictoryAxis,
  VictoryChart,
  VictoryLegend,
  VictoryLine,
  VictoryScatter,
  VictoryTheme,
} from 'victory-native'

import { strings } from '@/locales'
import { hourStringFromISODate } from '@/services'
import { Theme } from '@/ui'
import { Interval, TemperatureSeries } from './types'

type Props = {
  interval: Interval
}

export const TemperatureChart: React.FC<Props> = ({ interval }) => {
  const { colors, spacing } = useTheme<Theme>()

  const bedTempData = formatTemperatureSeries(interval.timeseries.tempBedC)
  const roomTempData = formatTemperatureSeries(interval.timeseries.tempRoomC)

  const chartLegends = [
    {
      name: strings.sleepDetails.bed,
      symbol: { fill: colors.blue },
    },
    {
      name: strings.sleepDetails.room,
      symbol: { fill: colors.green },
    },
  ]

  return (
    <VictoryChart
      animate={{ duration: 500 }}
      height={250}
      padding={{ top: 20, bottom: 30, left: 50, right: 50 }}
      scale={{ x: 'time', y: 'linear' }}
      theme={VictoryTheme.material}
    >
      <VictoryAxis
        dependentAxis
        domain={{ y: [15, 40] }}
        tickFormat={(t) => `${t} ${strings.common.celsius}`}
        style={{
          axis: {
            stroke: colors.text,
          },
          grid: {
            stroke: 'none',
          },
          ticks: {
            stroke: 'none',
          },
          tickLabels: {
            fill: colors.text,
          },
        }}
      />
      <VictoryAxis
        fixLabelOverlap
        style={{
          axis: {
            stroke: colors.text,
          },
          grid: {
            stroke: 'none',
          },
          ticks: {
            stroke: 'none',
          },
          tickLabels: {
            fill: colors.text,
          },
        }}
        tickFormat={(date: Date) => hourStringFromISODate(date)}
      />
      <VictoryLegend
        x={125}
        y={5}
        data={chartLegends}
        gutter={spacing.l}
        orientation="horizontal"
        style={{
          labels: {
            fill: colors.text,
            fillOpacity: 0.8,
          },
        }}
      />
      <VictoryLine
        data={bedTempData}
        interpolation="natural"
        style={{
          data: {
            stroke: colors.blue,
          },
        }}
      />
      <VictoryLine
        data={roomTempData}
        interpolation="natural"
        style={{
          data: {
            stroke: colors.green,
          },
        }}
      />
      <VictoryScatter
        animate={{
          onLoad: {
            before: () => ({ opacity: 0 }),
            after: () => ({ opacity: 1 }),
          },
        }}
        data={bedTempData}
        labels={({ datum }) => datum.y.toFixed(1)}
        style={{
          data: {
            fill: colors.blue,
            opacity: ({ datum }) => datum.opacity,
          },
          labels: {
            fill: colors.blue,
            fontSize: 10,
            opacity: ({ datum }) => datum.opacity,
          },
        }}
      />
      <VictoryScatter
        animate={{
          onLoad: {
            before: () => ({ opacity: 0 }),
            after: () => ({ opacity: 1 }),
          },
        }}
        data={roomTempData}
        labels={({ datum }) => datum.y.toFixed(1)}
        style={{
          data: {
            fill: colors.green,
            opacity: ({ datum }) => datum.opacity,
          },
          labels: {
            fill: colors.green,
            fontSize: 10,
            opacity: ({ datum }) => datum.opacity,
          },
        }}
      />
    </VictoryChart>
  )
}

function formatTemperatureSeries(tempSeries: TemperatureSeries) {
  return tempSeries.map(([date, temp]) => ({
    x: new Date(date),
    y: temp,
  }))
}
