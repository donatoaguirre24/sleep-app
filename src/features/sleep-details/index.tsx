import { useTheme } from '@shopify/restyle'
import React, { useLayoutEffect } from 'react'
import { ScrollView } from 'react-native'

import { subjectData } from '@/data'
import { strings } from '@/locales'
import { SleepDetailsStackProps } from '@/navigation'
import { dayStringFromISODate } from '@/services'
import { Box, Text, Theme } from '@/ui'
import { HealthMetricCard } from './health-metric-card'
import { SleepStagesChart } from './sleep-stages-chart'
import { TemperatureChart } from './temperature-chart'

export const SleepDetails: React.FC<SleepDetailsStackProps> = ({ navigation, route }) => {
  const { spacing } = useTheme<Theme>()

  const { setOptions } = navigation
  const { intervalId, name, userId, averageHeartRate, averageRespiratoryRate } = route.params

  const { intervals } = subjectData[userId]
  const interval = intervals.find(({ id }) => id === intervalId)! // Non-Null assertion

  useLayoutEffect(() => {
    setOptions({ title: `${dayStringFromISODate(interval.ts)} - ${name}` })
  }, [setOptions, interval.ts, name])

  return (
    <ScrollView contentContainerStyle={{ paddingHorizontal: spacing.m, paddingBottom: spacing.xl }}>
      <Box marginBottom="m">
        <Text variant="subheader" marginBottom="m">
          {strings.sleepDetails.healthMetricsTitle}
        </Text>
        <Box flexDirection="row" paddingHorizontal="s">
          <HealthMetricCard
            flex={1}
            title={strings.sleepDetails.heartRateTitle}
            metric={averageHeartRate}
            unit={`${strings.common.bpm} (${strings.common.averageShort})`}
            color="magenta"
          />
          <Box marginHorizontal="s" />
          <HealthMetricCard
            flex={1}
            title={strings.sleepDetails.respiratoryRateTitle}
            metric={averageRespiratoryRate}
            unit={`${strings.common.BrPM} (${strings.common.averageShort})`}
            color="blue"
          />
        </Box>
      </Box>
      <Box marginBottom="m">
        <Text variant="subheader" marginBottom="xs">
          {strings.sleepDetails.stagesChartTitle}
        </Text>
        <SleepStagesChart interval={interval} />
      </Box>
      <Box>
        <Text variant="subheader" marginBottom="xs">
          {strings.sleepDetails.temperatureChartTitle}
        </Text>
        <TemperatureChart interval={interval} />
      </Box>
    </ScrollView>
  )
}
