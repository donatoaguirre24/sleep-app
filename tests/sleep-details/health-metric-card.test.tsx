import { ResponsiveValue } from '@shopify/restyle'
import '@testing-library/jest-native'
import { render, screen } from '@testing-library/react-native'
import React from 'react'

import { HealthMetricCard } from '@/features/sleep-details/health-metric-card'
import { Theme } from '@/ui'
import { withProviders } from '../utils'

describe('HealthMetricCard', () => {
  it('renders correctly', () => {
    render(withProviders(<HealthMetricCard {...mockData} />))

    expect(screen.toJSON()).toMatchSnapshot()
  })

  it('should render the metric and unit properly', () => {
    render(withProviders(<HealthMetricCard {...mockData} testID="card" />))

    const card = screen.getByTestId('card')

    expect(card).toHaveTextContent('60 bpm (avg)')
  })

  it('should have the correct title color', () => {
    render(withProviders(<HealthMetricCard {...mockData} />))

    const title = screen.getByText('Heart Rate')

    expect(title).toHaveStyle({ color: '#FF006A' })
  })
})

const mockData = {
  title: 'Heart Rate',
  metric: '60',
  unit: 'bpm (avg)',
  color: 'magenta' as ResponsiveValue<'purple', Theme>,
}
