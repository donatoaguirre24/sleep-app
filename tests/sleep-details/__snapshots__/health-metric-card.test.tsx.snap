// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HealthMetricCard renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "minHeight": 75,
        },
        Array [
          Object {
            "borderColor": "#CCCCCC",
            "borderRadius": 8,
            "borderWidth": 0,
            "padding": 8,
            "shadowColor": "#CCCCCC",
            "shadowOffset": Object {
              "height": 3,
              "width": 1,
            },
            "shadowOpacity": 0.5,
            "shadowRadius": 4,
          },
        ],
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FF006A",
            "fontSize": 16,
            "fontWeight": "bold",
            "lineHeight": 18,
          },
        ]
      }
    >
      Heart Rate
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#1C1C1E",
              "fontSize": 24,
              "fontWeight": "bold",
              "lineHeight": 30,
            },
          ]
        }
      >
        60
         
        <Text
          style={
            Array [
              Object {
                "color": "#5E5E5E",
                "fontSize": 16,
                "fontWeight": "bold",
                "lineHeight": 18,
              },
            ]
          }
        >
          bpm (avg)
        </Text>
      </Text>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
